app/Domains/RevenueStreams/Actions/CreateRevenueStream.php

```diff
@@ -29,14 +31,23 @@ public function handle(array $params): RevenueStream
         return RevenueStream::create($params);
     }
 
-    public function rules(): array
+    public function rules(ActionRequest $request): array
     {
-        return [
+        $rules = [
             'name' => ['required', 'string', 'max:255'],
             'description' => ['nullable', 'string'],
             'type_id' => ['required', 'exists:revenue_stream_types,id'],
-            'values' => ['required', 'json'],
+            'values' => ['required', 'array'],
         ];
+
+        $type = RevenueStreamType::find($request->type_id);
+        if ($type) {
+            foreach ($type->properties as $property) {
+                $rules['values.' . $property['name']] = ['required', Rule::in(['single_line', 'multi_line', 'text', 'range', 'radio', 'checkbox', 'dropbox', 'repeater'])];
```

Comment:
Use the `RevenueStreamTypeEnum` here.
Also, please check the [comment on the `RevenueStreamType` PR](https://github.com/Eveneer/evengoams-backend/pull/62#pullrequestreview-2228760683)